File: backend/app/schemas/deal.py
"""
Deal schemas for request/response validation.
"""
from typing import Optional, List, Any, Dict
from pydantic import BaseModel, Field
from decimal import Decimal


class DealBase(BaseModel):
    """Base deal schema with common fields."""
    
    Deal_Name: str = Field(..., description="Deal name (required)")
    Account_Name: Optional[Dict[str, Any]] = Field(None, description="Associated account")
    Contact_Name: Optional[Dict[str, Any]] = Field(None, description="Associated contact")
    Amount: Optional[float] = Field(None, description="Deal amount")
    Stage: Optional[str] = Field(None, description="Deal stage")
    Closing_Date: Optional[str] = Field(None, description="Expected closing date")
    Probability: Optional[int] = Field(None, description="Probability of closing (%)")
    Type: Optional[str] = Field(None, description="Deal type")
    Lead_Source: Optional[str] = Field(None, description="Lead source")
    Description: Optional[str] = Field(None, description="Deal description")
    
    # Custom fields for TBDC
    Industry: Optional[str] = Field(None, description="Industry vertical")
    Company_Website: Optional[str] = Field(None, description="Company website URL")
    Support_Required: Optional[str] = Field(None, description="Support required from TBDC")


class DealCreate(DealBase):
    """Schema for creating a new deal."""
    pass


class DealUpdate(BaseModel):
    """Schema for updating a deal (all fields optional)."""
    
    Deal_Name: Optional[str] = None
    Account_Name: Optional[Dict[str, Any]] = None
    Contact_Name: Optional[Dict[str, Any]] = None
    Amount: Optional[float] = None
    Stage: Optional[str] = None
    Closing_Date: Optional[str] = None
    Probability: Optional[int] = None
    Type: Optional[str] = None
    Lead_Source: Optional[str] = None
    Description: Optional[str] = None
    Industry: Optional[str] = None
    Company_Website: Optional[str] = None
    Support_Required: Optional[str] = None


class DealResponse(BaseModel):
    """Response schema for a single deal."""
    
    data: Dict[str, Any] = Field(..., description="Deal data from Zoho")
    
    class Config:
        extra = "allow"


class DealListResponse(BaseModel):
    """Response schema for a list of deals."""
    
    data: List[Dict[str, Any]] = Field(default_factory=list, description="List of deals")
    page: int = Field(1, description="Current page number")
    per_page: int = Field(50, description="Records per page")
    total_count: int = Field(0, description="Total number of records in this response")
    more_records: bool = Field(False, description="Whether more records are available")

File: frontend/src/types/deal.ts
// Revenue customer information
export interface RevenueCustomer {
  name: string;
  industry: string;
  revenue_contribution: string;
  description: string;
}

// Pricing line item for recommended services
export interface PricingLineItem {
  service_name: string;
  description: string;
  category: string; // "core_service" | "customer_meeting" | "investor_meeting" | "additional_service"
  quantity: number;
  unit_price_eur: number;
  total_price_eur: number;
}

// Pricing summary with recommended services and total cost
export interface PricingSummary {
  recommended_services: PricingLineItem[];
  total_cost_eur: number;
  pricing_notes: string[];
}

// AI-generated analysis for a deal
export interface DealAnalysis {
  company_name?: string;
  country?: string;
  region?: string;
  summary?: string;
  product_description?: string;
  vertical?: string;
  business_model?: string;
  motion?: string;
  raise_stage?: string;
  company_size?: string;
  // Revenue & Customers
  revenue_top_5_customers?: RevenueCustomer[];
  // Scoring
  scoring_rubric?: Record<string, unknown>;
  fit_score?: number;
  fit_assessment?: string;
  // ICP & Support
  icp_mapping?: string;
  likely_icp_canada?: string;
  support_required?: string;
  support_recommendations?: string[];
  // Pricing
  pricing_summary?: PricingSummary;
  // Insights
  key_insights?: string[];
  questions_to_ask?: string[];
  confidence_level?: string;
  notes?: string[];
}

// Marketing material matched by semantic similarity
export interface MarketingMaterial {
  material_id: string;
  title: string;
  link: string;
  industry?: string;
  business_topics?: string;
  similarity_score?: number;
}

// Similar customer identified by LLM analysis
export interface SimilarCustomer {
  name: string;
  description: string;
  industry?: string;
  website?: string;
  why_similar?: string;
}

export interface Deal {
  id: string;
  Deal_Name: string;
  Account_Name?: {
    id: string;
    name: string;
  };
  Contact_Name?: {
    id: string;
    name: string;
  };
  Amount?: number;
  Stage?: string;
  Closing_Date?: string;
  Probability?: number;
  Type?: string;
  Lead_Source?: string;
  Description?: string;
  Created_Time?: string;
  Modified_Time?: string;
  Owner?: {
    id: string;
    name: string;
    email: string;
  };
  // Custom fields
  Industry?: string;
  Company_Website?: string;
  Support_Required?: string;
  // Allow for any additional Zoho fields
  [key: string]: unknown;
}

// Deal with analysis data (used when fetching single deal details)
export interface DealWithAnalysis {
  data: Deal;
  analysis?: DealAnalysis;
  analysis_available?: boolean;
  from_cache?: boolean;
  marketing_materials?: MarketingMaterial[];
  similar_customers?: SimilarCustomer[];
}

export interface DealListResponse {
  data: Deal[];
  page: number;
  per_page: number;
  total_count: number;
  more_records: boolean;
}

export interface DealResponse {
  data: Deal;
  analysis?: DealAnalysis;
  analysis_available?: boolean;
  from_cache?: boolean;
  marketing_materials?: MarketingMaterial[];
  similar_customers?: SimilarCustomer[];
}
